apps/emqx/src/emqx_router_helper.erl

```diff
@@ -188,6 +190,169 @@ code_change(_OldVsn, State, _Extra) ->
+            schedule_task({purge, Node}, 0);
+        false ->
+            false
+    end.
+
+schedule_purge_left(Node) ->
+    case am_core() andalso pick_responsible({purge, Node}) of
+        Responsible when Responsible == node() ->
+            %% Schedule purge on responsible node first, to avoid racing for a global lock.
+            schedule_task({purge, Node}, 0);
+        _ ->
+            %% Replicant / not responsible.
+            %% In the latter case try to purge on the next reconcile, as a fallback.
+            false
+    end.
+
+handle_purge(Node, State) ->
+    try purge_dead_node_trans(Node) of
+        true ->
+            ?tp(debug, emqx_router_node_purged, #{node => Node}),
+            forget_node(Node, State);
+        false ->
+            ?tp(debug, emqx_router_node_purge_skipped, #{node => Node}),
+            forget_node(Node, State);
+        aborted ->
+            ?tp(notice, emqx_router_node_purge_aborted, #{node => Node}),
+            State
+    catch
+        Kind:Error ->
+            ?tp(warning, emqx_router_node_purge_error, #{
+                node => Node,
+                kind => Kind,
+                error => Error
+            }),
+            State
+    end.
+
+purge_dead_node(Node) ->
+    case node_has_routes(Node) of
+        true ->
+            ok = cleanup_routes(Node),
+            true;
+        false ->
+            false
+    end.
+
+purge_dead_node_trans(Node) ->
+    case node_has_routes(Node) of
+        true ->
+            global:trans(
+                {?LOCK(Node), self()},
+                fun() ->
+                    ok = cleanup_routes(Node),
+                    true
+                end,
+                cores(),
+                _Retries = 3
+            );
+        false ->
+            false
+    end.
+
+purge_this_node() ->
+    %% TODO: Guard against possible `emqx_router_helper` restarts?
+    purge_dead_node(node()).
+
+node_has_routes(Node) ->
+    ets:member(?ROUTING_NODE, Node).
+
+cleanup_routes(Node) ->
+    emqx_router:cleanup_routes(Node),
+    mria:dirty_delete(?ROUTING_NODE, Node).
+
+%%
+
+schedule_task(Task, Timeout) ->
+    emqx_utils:start_timer(choose_timeout(Timeout), {start, Task}).
+
+handle_task({purge, Node}, State) ->
+    handle_purge(Node, State);
+handle_task(reconcile, State) ->
+    handle_reconcile(State).
+
+choose_timeout({Baseline, '±', Jitter}) ->
+    Baseline + rand:uniform(Jitter * 2) - Jitter;
+choose_timeout(Baseline) ->
+    Baseline.
+
+%%
+
+am_core() ->
+    mria_config:whoami() =/= replicant.
+
+cores() ->
+    %% Include stopped nodes as well.
+    mria_membership:nodelist().
+
+pick_responsible(Task) ->
+    %% Pick a responsible core node.
+    %% We expect the same node to be picked as responsible across the cluster (unless
+    %% the cluster is highly turbulent).
+    Nodes = lists:sort(mria_membership:running_core_nodelist()),
+    N = length(Nodes),
+    N > 0 andalso lists:nth(1 + erlang:phash2(Task, N), Nodes).
```

Comment:
nit: the return value calculation and its matching was a bit difficult to understand